File: tools/math_dsl/cascade.py
"""W244 wave 20 — closed-form analytical model for `cascade` (tumble) feature.

Industry pattern (Pragmatic Sweet Bonanza, Relax Money Train, Play'n GO
Reactoonz, BTG Bonanza, all tumble/avalanche/cascade games):

  Cascade dynamics
  ----------------
    On win: winning symbols disappear, remaining symbols fall down,
    new symbols drop from top. If new arrangement triggers another
    win, cascade continues. Each consecutive cascade may carry an
    escalating multiplier (e.g. 1× → 2× → 4× → 8× → 16×).

  Closed-form RTP contribution
  ----------------------------
    Per-spin RTP from cascade tail (BEYOND the first win):

      Let p_win_per_cascade = P(cascade step also produces a win)
          base_pay_per_cascade = E[pay × bet | cascade step won]
          multiplier_at_step_n = `multiplier_ladder[n]` (1-indexed)

      E[total_pay_per_trigger] = sum_{n=1..max_chain} (
        P(reach cascade step n) × base_pay_per_cascade × multiplier_at_step_n
      )

      where P(reach step n) = p_initial_win × p_win_per_cascade^(n-1)

    Geometric chain with multiplier ramp. Bounded by `max_chain` cap.

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_cascade_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_cascade_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class CascadeParams:
    """Closed-form model inputs."""
    p_initial_win: float                    # P(base spin produces a win)
    base_pay_per_cascade_x_bet: float       # E[pay × bet | cascade won]
    p_win_per_cascade: float                # P(cascade step also wins)
    multiplier_ladder: tuple[float, ...]    # mult at step 1, 2, 3, ...
    max_chain: int = 16                     # absolute cap on cascade length

    def __post_init__(self):
        if not (0.0 <= self.p_initial_win <= 1.0):
            raise ValueError(
                f"p_initial_win {self.p_initial_win} outside [0,1]"
            )
        if self.base_pay_per_cascade_x_bet < 0:
            raise ValueError("base_pay_per_cascade_x_bet must be ≥ 0")
        if not (0.0 <= self.p_win_per_cascade <= 1.0):
            raise ValueError(
                f"p_win_per_cascade {self.p_win_per_cascade} outside [0,1]"
            )
        if not self.multiplier_ladder:
            raise ValueError("multiplier_ladder must be non-empty")
        if any(m < 0 for m in self.multiplier_ladder):
            raise ValueError("multiplier_ladder entries must be ≥ 0")
        if self.max_chain < 1:
            raise ValueError("max_chain must be ≥ 1")


def expected_chain_length(params: CascadeParams) -> float:
    """E[number of cascade steps after the initial win | initial win triggered].

    Geometric arrival truncated at max_chain. For p_win_per_cascade < 1:

      E[k_cascades] = sum_{k=1..max_chain} P(at least k cascades)
                    = sum_{k=1..max_chain} p_win_per_cascade^k
                    = p × (1 - p^max_chain) / (1 - p)   if p < 1

    For p = 1: deterministic max_chain.
    """
    p = params.p_win_per_cascade
    n = params.max_chain
    if p >= 1.0:
        return float(n)
    if p <= 0:
        return 0.0
    return p * (1.0 - p ** n) / (1.0 - p)


def expected_pay_per_trigger(params: CascadeParams) -> float:
    """E[total pay × bet | initial cascade triggered].

    sum_{n=1..max_chain} P(reach step n) × base_pay × multiplier_at_step_n

    where P(reach step n) = p_win_per_cascade^(n-1) (n=1 is the initial win).
    """
    total = 0.0
    p_chain = 1.0  # P(reach step 1 | initial win triggered) = 1
    for step in range(1, params.max_chain + 1):
        idx = min(step - 1, len(params.multiplier_ladder) - 1)
        mult = params.multiplier_ladder[idx]
        total += p_chain * params.base_pay_per_cascade_x_bet * mult
        # Advance to next step: multiply by p_win_per_cascade
        p_chain *= params.p_win_per_cascade
    return total


def cascade_rtp(params: CascadeParams) -> dict:
    """Per-base-spin RTP contribution + audit breakdown."""
    e_chain_len = expected_chain_length(params)
    e_pay_per_trigger = expected_pay_per_trigger(params)
    rtp = params.p_initial_win * e_pay_per_trigger

    # Per-step expected contribution breakdown
    per_step = []
    p_chain = 1.0
    for step in range(1, params.max_chain + 1):
        idx = min(step - 1, len(params.multiplier_ladder) - 1)
        mult = params.multiplier_ladder[idx]
        contrib = p_chain * params.base_pay_per_cascade_x_bet * mult
        per_step.append({
            "step": step,
            "p_reach": p_chain,
            "multiplier": mult,
            "contribution_x_bet": contrib,
        })
        p_chain *= params.p_win_per_cascade

    return {
        "rtp_contribution": rtp,
        "p_initial_win": params.p_initial_win,
        "base_pay_per_cascade_x_bet": params.base_pay_per_cascade_x_bet,
        "p_win_per_cascade": params.p_win_per_cascade,
        "max_chain": params.max_chain,
        "expected_chain_length": e_chain_len,
        "expected_pay_per_trigger_x_bet": e_pay_per_trigger,
        "per_step_breakdown": per_step,
    }

File: rust-sim/src/kernels/cascade.rs
//! W244.20 — cascade Rust port.
//!
//! Closed-form: bounded geometric chain × multiplier_ladder per step.
//!   RTP = p_initial_win × sum_{n=1..N}(p^(n-1) × base × mult[n])

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CascadeParams {
    pub p_initial_win: f64,
    pub base_pay_per_cascade_x_bet: f64,
    pub p_win_per_cascade: f64,
    pub multiplier_ladder: Vec<f64>,
    #[serde(default = "default_max_chain")]
    pub max_chain: u32,
}

fn default_max_chain() -> u32 { 16 }

impl CascadeParams {
    pub fn validate(&self) -> Result<(), String> {
        if !(0.0..=1.0).contains(&self.p_initial_win) {
            return Err(format!("p_initial_win {} outside [0,1]", self.p_initial_win));
        }
        if self.base_pay_per_cascade_x_bet < 0.0 {
            return Err("base_pay_per_cascade_x_bet must be ≥ 0".into());
        }
        if !(0.0..=1.0).contains(&self.p_win_per_cascade) {
            return Err(format!("p_win_per_cascade {} outside [0,1]", self.p_win_per_cascade));
        }
        if self.multiplier_ladder.is_empty() {
            return Err("multiplier_ladder must be non-empty".into());
        }
        for &m in &self.multiplier_ladder {
            if m < 0.0 {
                return Err("multiplier_ladder entries must be ≥ 0".into());
            }
        }
        if self.max_chain < 1 {
            return Err("max_chain must be ≥ 1".into());
        }
        Ok(())
    }
}

pub fn expected_chain_length(p: &CascadeParams) -> f64 {
    let pp = p.p_win_per_cascade;
    let n = p.max_chain as i32;
    if pp >= 1.0 {
        return n as f64;
    }
    if pp <= 0.0 {
        return 0.0;
    }
    pp * (1.0 - pp.powi(n)) / (1.0 - pp)
}

pub fn expected_pay_per_trigger(p: &CascadeParams) -> f64 {
    let mut total = 0.0_f64;
    let mut p_chain = 1.0_f64;
    let ladder_len = p.multiplier_ladder.len();
    for step in 1..=p.max_chain {
        let idx = (step as usize - 1).min(ladder_len - 1);
        let mult = p.multiplier_ladder[idx];
        total += p_chain * p.base_pay_per_cascade_x_bet * mult;
        p_chain *= p.p_win_per_cascade;
    }
    total
}

#[derive(Debug, Serialize)]
pub struct PerStep {
    pub step: u32,
    pub p_reach: f64,
    pub multiplier: f64,
    pub contribution_x_bet: f64,
}

#[derive(Debug, Serialize)]
pub struct CascadeResult {
    pub rtp_contribution: f64,
    pub p_initial_win: f64,
    pub base_pay_per_cascade_x_bet: f64,
    pub p_win_per_cascade: f64,
    pub max_chain: u32,
    pub expected_chain_length: f64,
    pub expected_pay_per_trigger_x_bet: f64,
    pub per_step_breakdown: Vec<PerStep>,
}

pub fn cascade_rtp(p: &CascadeParams) -> CascadeResult {
    let e_chain_len = expected_chain_length(p);
    let e_pay = expected_pay_per_trigger(p);
    let rtp = p.p_initial_win * e_pay;

    let mut per_step: Vec<PerStep> = Vec::with_capacity(p.max_chain as usize);
    let mut p_chain = 1.0_f64;
    let ladder_len = p.multiplier_ladder.len();
    for step in 1..=p.max_chain {
        let idx = (step as usize - 1).min(ladder_len - 1);
        let mult = p.multiplier_ladder[idx];
        per_step.push(PerStep {
            step,
            p_reach: p_chain,
            multiplier: mult,
            contribution_x_bet: p_chain * p.base_pay_per_cascade_x_bet * mult,
        });
        p_chain *= p.p_win_per_cascade;
    }

    CascadeResult {
        rtp_contribution: rtp,
        p_initial_win: p.p_initial_win,
        base_pay_per_cascade_x_bet: p.base_pay_per_cascade_x_bet,
        p_win_per_cascade: p.p_win_per_cascade,
        max_chain: p.max_chain,
        expected_chain_length: e_chain_len,
        expected_pay_per_trigger_x_bet: e_pay,
        per_step_breakdown: per_step,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_cascade_first_step_only() {
        let p = CascadeParams {
            p_initial_win: 0.1,
            base_pay_per_cascade_x_bet: 1.0,
            p_win_per_cascade: 0.0,
            multiplier_ladder: vec![1.0],
            max_chain: 5,
        };
        // Step 1: 1.0 × 1.0 × 1.0 = 1.0
        assert!((expected_pay_per_trigger(&p) - 1.0).abs() < 1e-12);
    }

    #[test]
    fn geometric_5_step_doubling() {
        let p = CascadeParams {
            p_initial_win: 0.1,
            base_pay_per_cascade_x_bet: 1.0,
            p_win_per_cascade: 0.5,
            multiplier_ladder: vec![1.0, 2.0, 4.0, 8.0, 16.0],
            max_chain: 5,
        };
        // 1×1 + 0.5×2 + 0.25×4 + 0.125×8 + 0.0625×16 = 5.0
        assert!((expected_pay_per_trigger(&p) - 5.0).abs() < 1e-10);
    }

    #[test]
    fn p_zero_chain_length_zero() {
        let p = CascadeParams {
            p_initial_win: 0.3,
            base_pay_per_cascade_x_bet: 1.0,
            p_win_per_cascade: 0.0,
            multiplier_ladder: vec![1.0],
            max_chain: 5,
        };
        assert_eq!(expected_chain_length(&p), 0.0);
    }
}
